File: luaui/geom/Rectangle.lua
local Rectangle = {}
Rectangle.__index = Rectangle

function Rectangle.new(left, top, width, height)
    local self = setmetatable({}, Rectangle)
    self.left = left or 0
    self.top = top or 0
    self.width = width or 0
    self.height = height or 0
    return self
end

function Rectangle:set(left, top, width, height)
    self.left, self.top, self.width, self.height = left or 0, top or 0, width or 0, height or 0
end

function Rectangle:get_right() return self.left + self.width end
function Rectangle:get_bottom() return self.top + self.height end

function Rectangle:intersection(rect, noAlloc)
    if noAlloc == nil then noAlloc = true end
    if noAlloc and self._intersectionCache == nil then self._intersectionCache = Rectangle.new() end

    local x0 = math.max(self.left, rect.left)
    local x1 = math.min(self:get_right(), rect:get_right())
    if x1 <= x0 then
        local r = noAlloc and self._intersectionCache or Rectangle.new()
        r:set(0, 0, 0, 0)
        return r
    end

    local y0 = math.max(self.top, rect.top)
    local y1 = math.min(self:get_bottom(), rect:get_bottom())
    if y1 <= y0 then
        local r = noAlloc and self._intersectionCache or Rectangle.new()
        r:set(0, 0, 0, 0)
        return r
    end

    local r = noAlloc and self._intersectionCache or Rectangle.new()
    r:set(x0, y0, x1 - x0, y1 - y0)
    return r
end

return Rectangle

File: flutter_ui/lib/geom/rectangle.dart
class Rectangle {
  double left, top, width, height;

  Rectangle([this.left = 0, this.top = 0, this.width = 0, this.height = 0]);

  void set(
      [double left = 0, double top = 0, double width = 0, double height = 0]) {
    this.left = left;
    this.top = top;
    this.width = width;
    this.height = height;
  }

  double get right => left + width;
  set right(double value) => width = value - left;

  double get bottom => top + height;
  set bottom(double value) => height = value - top;

  void inflate(double dx, double dy) {
    left -= dx;
    width += dx * 2;
    top -= dy;
    height += dy * 2;
  }

  bool equals(Rectangle? rc) {
    if (rc == null) return false;
    return rc.left == left &&
        rc.top == top &&
        rc.width == width &&
        rc.height == height;
  }

  bool containsPoint(double x, double y) =>
      x >= left && x < left + width && y >= top && y < top + height;

  bool containsRect(Rectangle rect) {
    if (rect.width <= 0 || rect.height <= 0) {
      return rect.left > left &&
          rect.top > top &&
          rect.right < right &&
          rect.bottom < bottom;
    } else {
      return rect.left >= left &&
          rect.top >= top &&
          rect.right <= right &&
          rect.bottom <= bottom;
    }
  }

  bool intersects(Rectangle rect) {
    double x0 = left < rect.left ? rect.left : left;
    double x1 = right > rect.right ? rect.right : right;
    if (x1 <= x0) return false;
    double y0 = top < rect.top ? rect.top : top;
    double y1 = bottom > rect.bottom ? rect.bottom : bottom;
    return y1 > y0;
  }

  Rectangle? _intersectionCache;
  Rectangle intersection(Rectangle rect, {bool noAlloc = true}) {
    if (noAlloc && _intersectionCache == null) _intersectionCache = Rectangle();
    double x0 = left < rect.left ? rect.left : left;
    double x1 = right > rect.right ? rect.right : right;
    if (x1 <= x0) {
      if (noAlloc) {
        _intersectionCache!.set();
        return _intersectionCache!;
      }
      return Rectangle();
    }
    double y0 = top < rect.top ? rect.top : top;
    double y1 = bottom > rect.bottom ? rect.bottom : bottom;
    if (y1 <= y0) {
      if (noAlloc) {
        _intersectionCache!.set();
        return _intersectionCache!;
      }
      return Rectangle();
    }
    Rectangle r = noAlloc ? _intersectionCache! : Rectangle();
    r.set(x0, y0, x1 - x0, y1 - y0);
    return r;
  }

  Rectangle copy() => Rectangle(left, top, width, height);

  @override
  String toString() =>
      "{left: $left, top: $top, bottom: $bottom, right: $right, width: $width, height: $height}";
}
